rholang/src/main/scala/coop/rchain/rholang/interpreter/ChargingReducer.scala

```diff
@@ -31,13 +31,13 @@ object ChargingReducer {
 
   implicit def chargingReducer[F[_]](
       implicit reducer: Reduce[F],
-      costAccounting: CostAccounting[F]
+      cost: _cost[F]
   ): ChargingReducer[F] = new ChargingReducer[F] {
     def phlo: F[Cost] =
-      costAccounting.get()
+      cost.get
 
     def setPhlo(limit: Cost): F[Unit] =
-      costAccounting.set(limit)
+      cost.set(limit)
```

Comment:
AFAIK in the Tagless Final world this is sometimes called "vertical stacking of interpreters" (vs "horizontal" as in `F[_]: Log: Console`). https://github.com/SystemFw/Befunge-93/blob/master/src/main/scala/semantics/language.scala#L16

Reply:
No matter how you call it, it's a useless instance of what you're calling it - sorry :) The example you're referring to provides substantial abstraction and useful buildup without exposing what's used underneath to build it. What we have here is horizontal stacking in poor disguise :)

Reply:
You should be able to construct an instance of `Reducer` without worrying about cost accounting (if you do then you use `ChargingReducer` that adds this capability). You should be able to test `Reducer` functionality without costs and similarly you should be able to test cost accounting logic without knowing anything about `Par`s. That's btw the reason why there are `substitute*[charge]` methods. 

If all you ever do is interpret `Par`s and charge for them then why not fixate for `private val costLimit = AtomicInteger` that is reachable from everywhere and don't try to write generic interpreter? 

You've done great job with refactoring this mess and you should be able to finally build these abstractions (similarly to what can be found in the link) but you've decided to fix cost accounting and interpreter logic together.

Reply:
`able to construct an instance of Reducer without worrying about cost accounting` - Reducer is full of `charge` calls, I don't see what you're suggesting without further substantial refactoring, on a totally orthogonal axis

`test Reducer functionality without costs` - this should be achieved by providing a no-op cost account

`should be able to test cost accounting logic without knowing anything about Pars` - ideally, maybe. The cost should though be a function of (what data is being used, what aspect of its usage we charge for). A Par has different execution, substitution, and storage costs. So Chargable's current API is not enough for that, and at the same time it's obvious its only purpose is being able to Substitute[A] (it's not used elsewhere, effectively). I mean, maybe the intent was different - but intent doesn't compile, and for the abstraction to bring us any value it needs to be used instead of the concretion.

`the reason why there are substitute*[charge] methods.` - as noted, I understand that charging 'inside' Substitute[A] required a Chargable[A]. The problems I have with that is the above + what I mentioned in the other comment + it's being done on the surface of `Substitute` really - so it's cheaper (mental costs -wise) to do that outside of Substitute.

`why not fixate for private val costLimit = AtomicInteger` - we might do that, after having all usages hidden behind the `_cost[F]` capability.

`You've done great job with refactoring this mess and you should be able to finally build these abstractions (similarly to what can be found in the link) but you've decided to fix cost accounting and interpreter logic together.` - contrary to what one might think, the refactorings aren't being made for their own sake. The abstractions you mention, OTOH, don't solve any of our immediate goals, nor are they easily achievable. So we're taking a step 'back' in this axis, just to make a couple steps forward along the other axes, and then maybe retake a step towards an abstract 'charge reason' in a better way :)